File: Luogu/p3000-p3999/p3372/p3372.rs
#[warn(unused_variables)]
use std::io;

static mut A: [i64; 100010] = [0; 100010];
fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let mut s = input.split_whitespace();
    let n: i64 = s.next().unwrap().parse().unwrap();
    let m: i64 = s.next().unwrap().parse().unwrap();
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    unsafe {
        s = input.split_whitespace();
        for i in 1..=n {
            A[i as usize] = s.next().unwrap().parse().unwrap();
        }
    }
    unsafe {
        build(1, 1, n);
    }
    for _i in 0..m {
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        s = input.split_whitespace();
        let op: i64 = s.next().unwrap().parse().unwrap();
        let x: i64 = s.next().unwrap().parse().unwrap();
        let y: i64 = s.next().unwrap().parse().unwrap();
        if op == 1 {
            let k: i64 = s.next().unwrap().parse().unwrap();
            unsafe {
                segadd(1, x, y, k);
            }
        } else if op == 2 {
            let ans: i64 = unsafe { segsum(1, x, y) };
            println!("{}", ans);
        }
    }
}

static mut L: [i64; 800080] = [0; 800080];
static mut R: [i64; 800080] = [0; 800080];
static mut V: [i64; 800080] = [0; 800080];
static mut T: [i64; 800080] = [0; 800080];

unsafe fn pushup(p: usize) {
    V[p] = V[p << 1] + V[p << 1 | 1];
}
unsafe fn pushdown(p: usize) {
    if T[p] != 0 {
        V[p << 1] += (R[p << 1] - L[p << 1] + 1) * T[p];
        V[p << 1 | 1] += (R[p << 1 | 1] - L[p << 1 | 1] + 1) * T[p];
        T[p << 1] += T[p];
        T[p << 1 | 1] += T[p];
        T[p] = 0;
    }
}
unsafe fn build(p: usize, l: i64, r: i64) {
    L[p] = l;
    R[p] = r;
    if l == r {
        V[p] = A[l as usize];
        T[p] = 0;
        return;
    }
    let mid: i64 = (l + r) >> 1;
    build(p << 1, l, mid);
    build(p << 1 | 1, mid + 1, r);
    pushup(p);
}

unsafe fn segadd(p: usize, l: i64, r: i64, k: i64) {
    if L[p] >= l && R[p] <= r {
        V[p] += (R[p] - L[p] + 1) * k;
        T[p] += k;
        return;
    }
    pushdown(p);
    let mid: i64 = (L[p] + R[p]) >> 1;
    if l <= mid {
        segadd(p << 1, l, r, k)
    };
    if r > mid {
        segadd(p << 1 | 1, l, r, k)
    };
    pushup(p);
}
unsafe fn segsum(p: usize, l: i64, r: i64) -> i64 {
    if L[p] >= l && R[p] <= r {
        return V[p];
    };
    pushdown(p);
    let mut res: i64 = 0;
    let mid: i64 = (L[p] + R[p]) >> 1;
    if l <= mid {
        res += segsum(p << 1, l, r)
    };
    if r > mid {
        res += segsum(p << 1 | 1, l, r)
    };
    return res;
}

File: Luogu/p3000-p3999/p3372/p3372.cs
using System;

namespace Main
{
	public class SegmentTree
	{
		public class Node
		{
			public int l;
			public int r;
			public long sum;
			public long add;
			public Node()
			{
				l = 0;
				r = 0;
				sum = 0;
				add = 0;
			}
		}
		public Node[] tr;

		public void Pushup(int p)
		{
			tr[p].sum = tr[p << 1].sum + tr[p << 1 | 1].sum;
		}
		public void Pushdown(int p)
		{
			if (tr[p].add != 0)
			{
				tr[p << 1].add += tr[p].add;
				tr[p << 1 | 1].add += tr[p].add;
				tr[p << 1].sum += (tr[p << 1].r - tr[p << 1].l + 1) * tr[p].add;
				tr[p << 1 | 1].sum += (tr[p << 1 | 1].r - tr[p << 1 | 1].l + 1) * tr[p].add;
				tr[p].add = 0;
			}
		}

		public void Build(int p, int l, int r, long[] a)
		{
			tr[p].l = l;
			tr[p].r = r;
			tr[p].add = 0;
			if (l == r)
			{
				tr[p].sum = a[l];
				return;
			}
			int mid = (l + r) >> 1;
			Build(p << 1, l, mid, a);
			Build(p << 1 | 1, mid + 1, r, a);
			Pushup(p);
		}

		public void Segadd(int p, int l, int r, long k)
		{
			if (tr[p].l >= l && tr[p].r <= r)
			{
				tr[p].add += k;
				tr[p].sum += (tr[p].r - tr[p].l + 1) * k;
				return;
			}
			Pushdown(p);
			int mid = (tr[p].l + tr[p].r) >> 1;
			if (l <= mid) Segadd(p << 1, l, r, k);
			if (r > mid) Segadd(p << 1 | 1, l, r, k);
			Pushup(p);
		}
		public long Segsum(int p, int l, int r)
		{
			if (tr[p].l >= l && tr[p].r <= r)
			{
				return tr[p].sum;
			}
			Pushdown(p);
			long res = 0;
			int mid = (tr[p].l + tr[p].r) >> 1;
			if (l <= mid) res += Segsum(p << 1, l, r);
			if (r > mid) res += Segsum(p << 1 | 1, l, r);
			return res;
		}

		public SegmentTree(int n, long[] a)
		{
			tr = new Node[800010];
			for (int i = 0; i <= 800000; i++)
				tr[i] = new Node();
			Build(1, 1, n, a);
		}
	}
	class MainClass
	{
		static public int Read()
		{
			int x = 0;
			int c = Console.Read();
			while (c < '0' || c > '9') c = Console.Read();
			while (c >= '0' && c <= '9')
			{
				x = (x * 10) + (c - 48);
				c = Console.Read();
			}
			return x;
		}
		public static void Main(string[] args)
		{
			int n, m;
			n = Read();
			m = Read();
			long[] a = new long[100010];
			for (int i = 1; i <= n; i++)
				a[i] = Read();
			SegmentTree tr = new SegmentTree(n, a);
			for (int i = 1; i <= m; i++)
			{
				int op, l, r;
				op = Read();
				l = Read();
				r = Read();
				if (op == 1)
				{
					long k = Read();
					tr.Segadd(1, l, r, k);
				}
				else
				{
					Console.WriteLine(tr.Segsum(1, l, r));
				}
			}
		}
	}
}